fs/config/configfile/config_file.go

```diff
@@ -0,0 +1,60 @@
+package configfile
+
+import (
+	"io"
+
+	"github.com/Unknwon/goconfig"
+)
+
+type ConfigFile interface {
+	GetSectionList() []string
+	HasSection(section string) bool
+	DeleteSection(section string)
+
+	GetKeyList(section string) []string
+	GetValue(section string, key string) (string, error)
+	MustValue(section string, key string, default_ ...string) string
+	SetValue(section string, key string, value string)
+	DeleteKey(section string, key string) bool
+
+	Load(io.Reader) error
```

Comment:
I think having these take an `io.Reader` is probably an implementation detail, so maybe the interface should just be `Load() error) and `Save() error` - what do you think?

What is your plan for this? Do you want rclone to be able to encrypt its config? It doesn't seem quite flexible enough somehow.

Reply:
I added the `io.Reader` because with that I was able to not touch the encryption system. This would ofc. be a good opportunity to move that part too. Then I would add a NewGoConfig function which takes the reader and returns the ConfigFile

Reply:
> Yes currently the config source is hardcoded. I was just pushing an example to check if this is ok for you.

It is a great idea.

> I am thinking about setting the LoadConfig and SaveConfig functions to nil and then setting them to the GoConfig ones somewhere at start. This would completely remove any dependency to goconfig for external projects.

Hmm, I think maybe we are mixing up two possible concepts

1. The setting and getting of config data
2. The serializing of the config data into a config file.

I'm presuming that if you swap out 1 you'll provide your own implementation of 2 - is that your understanding?

I'm thinking that what you might do is use some kind of database for reading and writing the config. In which case the serialization into a config file is irrelevant and un-needed.

In which case all the saving and loading to files (including the encryption) should be in `fs/config/configfile` and the interface should just have a `Load() error` and `Save() error` in it which likely would be ignored by someone implementing their own config system.

What do you think?

Reply:
Sorry for not responding :) Didnt had much time lately. I think your idea is good, but I dont know yet how I would like to implement the registering of the custom config backend. I will play a bit with the source when I find time the next days

Reply:
I would like to have the encryption seperated from the config interface/implementation since the implementation is so generic that you can use it for other providers too.